chains/substrate/client/client.go

```diff
@@ -0,0 +1,126 @@
+// The Licensed Work is (c) 2022 Sygma
+// SPDX-License-Identifier: BUSL-1.1
+
+package client
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+
+	"github.com/ChainSafe/sygma-relayer/chains/substrate/connection"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/client"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/rpc/author"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/signature"
+	"github.com/centrifuge/go-substrate-rpc-client/v4/types"
+	"github.com/rs/zerolog/log"
+)
+
+type SubstrateClient struct {
+	client.Client
+	author.Author
+	key       *signature.KeyringPair // Keyring used for signing
+	ChainID   *big.Int
+	nonceLock sync.Mutex // Locks nonce for updates
+	nonce     types.U32  // Latest account nonce
+}
+
+func NewSubstrateClient(url string, key *signature.KeyringPair) (*SubstrateClient, error) {
+	c := &SubstrateClient{
+		key: key,
+	}
+	client, err := client.Connect(url)
+	if err != nil {
+		return nil, err
+	}
+	c.Client = client
+	return c, nil
+}
+
+// SendRawTransaction accepts rlp-encode of signed transaction and sends it via RPC call
+func (c *SubstrateClient) sendRawTransaction(ext types.Extrinsic) (types.Hash, error) {
+	return c.Author.SubmitExtrinsic(ext)
+}
+
+func (c *SubstrateClient) signAndSendTransaction(opts types.SignatureOptions, ext types.Extrinsic) (types.Hash, error) {
+
+	err := ext.Sign(*c.key, opts)
+	if err != nil {
+		c.nonceLock.Unlock()
```

Comment:
I don't understand how is `nonceLock` being used. Here you are calling `Unlock()` but shouldn't this mean that you initially locked nonce on entering function - maybe I am missing something?

Reply:
I was rushing, as I now see the entire context, but I still think this should be removed as you are unlocking nonce in line 119 after calling `signAndSendTransaction`